cranelift-wasm/src/code_translator.rs

```diff
@@ -987,18 +987,86 @@ pub fn translate_operator<FE: FuncEnvironment + ?Sized>(
         | Operator::Fence { .. } => {
             return Err(wasm_unsupported!("proposed thread operator {:?}", op));
         }
-        Operator::MemoryInit { .. }
-        | Operator::DataDrop { .. }
-        | Operator::MemoryCopy
-        | Operator::MemoryFill
-        | Operator::TableInit { .. }
-        | Operator::ElemDrop { .. }
-        | Operator::TableCopy
-        | Operator::TableGet { .. }
-        | Operator::TableSet { .. }
-        | Operator::TableGrow { .. }
-        | Operator::TableSize { .. } => {
-            return Err(wasm_unsupported!("proposed bulk memory operator {:?}", op));
+        Operator::MemoryCopy => {
+            // The WebAssembly MVP only supports one linear memory and we assume it here.
```

Comment:
Yeah, +1 for Benjamin's comment.  Or at least, ensure that we don't silently produce incorrect code as the result of this assumption.  Adding an assertion would cause the system to fail during compilation, which is ugly but effective.  Is it easy to instead cause a validation error to be propagated?  If if's really easy to do that, I'd vote for it instead.

Reply:
All operators's that have a reserved index in the binary format such as memory instructions (before multi-memory) and table instructions (before reftypes) have an error result in [wasmparser](https://github.com/bytecodealliance/wasmparser/blob/b3f6ed69039372f14ea940216bef43ee8f154f55/src/binary_reader.rs#L1274) if the index is not 0. So it should be safe to assume the index is 0 here. I'll update the comments to indicate this.

Reply:
Would that mean that if `wasmparser` changed behavior, non-zero indexes could silently make their way here? If so, wouldn't that be something we'd like to guard against?

Reply:
Yes, I just started looking over the patch again and there's some improvements we need to make.

`MemoryCopy`, `MemoryFill`, and `MemoryInit` don't encode a memory index and won't until multi-memory is supported. Once that happens we'll be forced by the compiler to update all of these sites, as wasmparser will start providing it as an immediate in the Operator enum. I don't think it's likely that wasmparser will start allowing non-zero memory indices but not start passing them along in the Operator enum.

However one change needed is that the normal load/store instructions do call `get_heap(MemoryIndex(0))?` which ensures that there is a memory in this module. We need to do that for these instructions as well.

`TableCopy` and `TableInit` are the same as above. Except that after [PR](https://github.com/bytecodealliance/wasmparser/pull/151) wasmparser will start giving us indices. Once that happens, we can update `make_table` to ensure that we only use 0 if reference types is not implemented.

Reply:
I mispoke, it looks like the `MemoryCopy...` instructions do call `get_heap`. Just the table instructions need to be updated.